### DISCUSSION TITLE
Invalid Metric Names with JDK 11?

### DISCUSSION THREAD

**Discussion by Rapierian** [open-telemetry/opentelemetry-java#6285] [Q&A]
We recently upgraded to JDK 11 and I'm noticing that the various CodeHeap metrics now have an `'` in the name, and so the otel agent is complaining when it instantiates the metrics instruments, i.e.:
`[otel.javaagent 2024-03-11 18:06:53:001 +0000] [main] WARN io.opentelemetry.sdk.metrics.SdkMeter - Instrument name "memory.pools.CodeHeap-'non-profiled-nmethods'.usage" is invalid, returning noop instrument. Instrument names must consist of 63 or fewer characters including alphanumeric, _, ., -, and start with a letter.
--`

I can't be the first person to encounter this, but I'm not seeing a solution anywhere for all of these warnings. What's the method for dealing with these invalid instrument names?

**Comment by jkwatson** [comment]
What's generating the metrics?

**Comment by Rapierian** [comment]
I think this is just a jvm metric that we're getting from auto-instrumentation. We do have these flags enabled so it could be one of these, however:

` ``OTEL_INSTRUMENTATION_AKKA_ACTOR_ENABLED: "true"
  OTEL_INSTRUMENTATION_AKKA_HTTP_ENABLED: "true"
  OTEL_INSTRUMENTATION_APACHE_HTTPASYNCCLIENT_ENABLED: "true"
  OTEL_INSTRUMENTATION_AWS_SDK_ENABLED: "true"
  OTEL_INSTRUMENTATION_DROPWIZARD_METRICS_ENABLED: "true"
  OTEL_INSTRUMENTATION_GUAVA_ENABLED: "true"
  OTEL_INSTRUMENTATION_HIKARICP_ENABLED: "true"
  OTEL_INSTRUMENTATION_JDBC_ENABLED: "true"
  OTEL_INSTRUMENTATION_NETTY_ENABLED: "true"
  OTEL_INSTRUMENTATION_PLAY_ENABLED: "true"
  OTEL_INSTRUMENTATION_PLAY_WS_ENABLED: "true"
  OTEL_INSTRUMENTATION_SCALA_FORK_JOIN_ENABLED: "true"```

**Comment by jack-berg** [comment]
Can you tell me more about your setup? What libraries / frameworks are you using? What version of the otel java agent?

The metric you're describing doesn't come from the standard [JVM runtime metrics](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/runtime/jvm-metrics.md) that the otel java agent provides. It may be coming from some instrumentation built directly into one of the libraries you depend on?

**Comment by Rapierian** [comment]
Hmm, that's interesting - because "CodeHeap 'non-profiled nmethods'" IS a JVM metric, it [looks like it was added as of JDK 9](https://julio-falbo.medium.com/understand-jvm-and-jit-compiler-part-4-9738194ad06e) and [gets enabled/disabled on these flags](https://docs.oracle.com/en/java/javase/17/vm/java-hotspot-virtual-machine-performance-enhancements.html#GUID-1D9B26AD-8E0A-4771-90DA-A81A2C1F5B55).

So I'm guessing one of the flags I've got above - probably the Dropwizard Metrics one - is taking that JVM metric and these similar ones with `'` and exposing them. Let me see if I can figure out which flag exposes that metric...
